packages/babel-parser/src/parser/expression.js

```diff
@@ -514,6 +517,29 @@ export default class ExpressionParser extends LValParser {
       this.next();
       expr = this.finishNode(node, "UpdateExpression");
     }
+
+    if (
+      isAwait &&
+      expr.type === "Identifier" &&
+      // RegExps are ambiguous, because they could be a division.
+      // ( and [ are also ambiguous.
+      (this.match(tt.name) ||
+        this.match(tt.num) ||
+        this.match(tt.bigint) ||
+        this.match(tt.string) ||
+        (this.state.type.keyword && !this.state.type.binop)) &&
+      !this.hasPrecedingLineBreak()
+    ) {
+      this.raise(
+        startPos,
+        this.hasPlugin("topLevelAwait")
+          ? "Await expressions are only allowed inside async functions or modules."
+          : "Await expressions are only allowed inside async functions.",
+      );
+
+      return this.parseAwait(startPos, startLoc);
```

Comment:
We can reset the `awaitPos` after await expression is parsed since we have thrown an recoverable error before.